proxylib/memcached/text/parser.go

```diff
@@ -0,0 +1,468 @@
+		rules = append(rules, &br)
+	}
+	return rules
+}
+
+// TextMemcacheParserFactory implements proxylib.ParserFactory
+type TextMemcacheParserFactory struct{}
+
+// Create creates binary memcached parser
+func (p *TextMemcacheParserFactory) Create(connection *proxylib.Connection) proxylib.Parser {
+	log.Infof("TextMemcacheParserFactory: Create: %v", connection)
+	return &TextMemcacheParser{connection: connection, replyQueue: make([]*replyIntent, 0)}
+}
+
+// compile time check for interface implementation
+var _ proxylib.ParserFactory = &TextMemcacheParserFactory{}
+
+var textMemcacheParserFactory *TextMemcacheParserFactory
+
+const (
+	parserName = "textmemcache"
+)
+
+func init() {
+	log.Info("init(): Registering textMemcacheParserFactory")
+	proxylib.RegisterParserFactory(parserName, textMemcacheParserFactory)
+	proxylib.RegisterL7RuleParser(parserName, L7TextMemcacheRuleParser)
+}
+
+// TextMemcacheParser implements proxylib.Parser
+type TextMemcacheParser struct {
+	connection *proxylib.Connection
+
+	requestCount uint32
+	replyCount   uint32
+	replyQueue   []*replyIntent
+}
+
+type replyIntent struct {
+	command   []byte
+	requestID uint32
+	denied    bool
+}
+
+var _ proxylib.Parser = &TextMemcacheParser{}
+
+// OnData parses binary memcached data
+func (p *TextMemcacheParser) OnData(reply, endStream bool, dataBuffers [][]byte, offset int) (proxylib.OpType, int) {
+	log.Infof("OnData with offset %d", offset)
+
+	if reply {
+		if p.injectFromQueue() {
+			return proxylib.INJECT, len(DeniedMsg)
+		}
+		if len(dataBuffers) == 0 {
+			return proxylib.NOP, 0
+		}
+	}
+
+	// TODO: don't copy data to new slices
+	data := (bytes.Join(dataBuffers, []byte{}))[offset:]
+	log.Infof("Data length: %d", len(data))
+
+	linefeed := bytes.Index(data, []byte("\r\n"))
+	if linefeed < 0 {
+		log.Infof("Did not receive full first line, asking for 1 byte more")
+		return proxylib.MORE, 1
+	}
+
+	// TODO: iterate over data without copying it to new slices
+	// Tokenizing in memcached is done by spaces: https://github.com/memcached/memcached/blob/master/memcached.c#L2978
+	tokens := bytes.Fields(data[:linefeed])
+
+	if !reply {
+		meta := textMemcacheMeta{
+			command: tokens[0],
+		}
+
+		frameLength := 0
+		noreply := false
+		if p.isCommandRetrieval(meta.command) {
+			// get, gets, gat, gats
+			if bytes.HasPrefix(meta.command, []byte("get")) {
+				meta.keys = tokens[1:]
+			} else if bytes.HasPrefix(meta.command, []byte("gat")) {
+				meta.keys = tokens[2:]
+			}
+			frameLength = linefeed + 2
+		} else if p.isCommandStorage(meta.command) {
+			// storage commands
+			meta.keys = tokens[1:2]
+			nBytes, err := strconv.Atoi(string(tokens[4]))
+			if err != nil {
+				log.Error("Failed to parse storage payload length")
+				return proxylib.ERROR, 0
+			}
+			// 2 additional bytes for terminating linefeed
+			frameLength = linefeed + 2 + nBytes + 2
+
+			if meta.command[0] == 'c' { //storage command is "cas"
+				noreply = len(tokens) == 7
```

Comment:
I'm sure this is in the spec somewhere but I don't understand the meaning of those magic numbers. Consider replacing the `7` with a constant string length: `len("special")`. The same goes for the 6,3,4,4 in the following lines.

Reply:
done, added named constants, as I am not sure what constant string length would have achieved.